src/common/UserAvatar.js

```diff
@@ -20,32 +24,36 @@ type Props = $ReadOnly<{|
  * @prop [children] - If provided, will render inside the component body.
  * @prop [onPress] - Event fired on pressing the component.
  */
-export default class UserAvatar extends PureComponent<Props> {
-  render() {
-    const { avatarUrl, children, size, onPress } = this.props;
-    const borderRadius = size / 8;
-    const style = {
-      height: size,
-      width: size,
-      borderRadius,
-    };
+const UserAvatar = function UserAvatar(props: Props) {
```

Comment:
> That arrangement actually seems slightly silly to me -- why not just write something like
> 
> ```js
> function UserAvatar(props: Props) { … }
> ```
> 
> in the first place? 😛

Ah, right—I left this out of my explanation; oops. I have never favored this form because, like `var`, it means I have to think about hoisting ([doc](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting)).

But I could be convinced that the trade-off is fine here; React component files aren't generally long (or shouldn't be, anyway). 🙂

Reply:
I see, that's helpful. It seems quite likely that's been other people's thinking too, and has helped motivate the `const foo = () => { … }` style.

I think the term "hoisting", which suggests an action of transforming the code into some other code, makes what's going on here more complicated than it needs to be. (You can see some ambivalence about it right at the top of that doc.)

With `var` there's some surprising semantics because you can have a declaration separate from an initialization, and "hoisting" may be a helpful idea for thinking about those semantics.

But with `function`, there's no such complication -- a function declaration `function foo() { … }` just means there's a function, and it's called `foo`, and you can call it from anywhere under that scope. If it's at module toplevel, you can call it from anywhere in the module.

There's really *less* to think about with this form than there is with `const foo = function() { … }` -- the latter form adds the complication that the function only exists in part of the code, namely after the declaration. So you have to put the code in a certain order, but only sometimes (only if you're referring to the function in code directly at top level, outside of any function... or in an IIFE... or more generally in code that can execute before the module's top-level code has completed execution), and you end up with lint rules like `no-use-before-define` that try to make things simple again by making the constraints tighter, but that make it impossible to write some kinds of code, or to write some other code naturally, and so we disable that rule in several files.

Whereas with a plain function declaration, you just have the function, and you can call it.

I agree that OTOH the complicated semantics of `var`, which "hoisting" is an attempt to explain, are a good reason to generally avoid using it.

Reply:
What do you think of this small tweak to one of their examples?

```
catName("Chloe");

const suffix = "the cat"

function catName(name) {
  console.log("My cat's name is " + name + " " + suffix);
}
```

Reply:
Hmm, that is indeed more complicated :slightly_smiling_face: .

I think from the perspective above, the thing that's making that example complicated isn't the `function` -- it's that there's nontrivial logic happening at toplevel. Because there's that `catName` call happening at toplevel, it depends on `suffix` being present, and so it depends on the `const suffix = …` having already happened -- IOW there's a required order between those two statements. There wouldn't be such an ordering constraint if the `catName` call were inside a function definition (in any form), and were only invoked via an export.

In general it's a bad practice to have substantial logic happening at module toplevel, because it means that requiring a module can have side effects, or at a minimum can lead to spending time doing computation. I think it's rare in our code -- the only frequent patterns of top-level code that come to mind, beyond function expressions, are a handful of higher-order functions like `connect` that do no visible work up front.

Reply:
> it's that there's nontrivial logic happening at toplevel

True, it might be happening at toplevel—but it might not, right;

```diff
class IosCompliantAppleAuthButton extends PureComponent<Props, State> {
  state = {
    isNativeButtonAvailable: undefined,
  };

  async componentDidMount() {
    this.setState({ isNativeButtonAvailable: await AppleAuthentication.isAvailableAsync() });
  }

  render() {
    const { style, onPress, theme } = this.props;
    const { isNativeButtonAvailable } = this.state;
+   
+   catName("Chloe");
+
+   const suffix = "the cat"
+   
+   function catName(name) {
+     console.log("My cat's name is " + name + " " + suffix);
+   }
+   
    // [etc.]
  }
}
```

Reply:
Yeah. And I'd look a lot more skeptically at a `function` declaration *inside* a function. I guess in fact one way to look at that is:
* the module toplevel is for definitions, and not logic -- so `function` statements are fine, and function expressions with long complex bodies (as in `const f = (…): … => { … }`)  are the bread and butter of our codebase, but nontrivial logic to be executed immediately requires care;
* a function body is for logic -- so `function` statements, and for that matter complex function expressions, make things complex and require care, but nontrivial logic is at home;
* and one reason that's good is that mixing definitions and logic at the same level leads to surprises like this one.

Reply:
Mmm, right; that's really helpful.